### ISSUE TITLE
Raise specific exception on traversing "uninitialised" relationships

### ISSUE THREAD

**Issue by aanastasiou** [neo4j-contrib/neomodel#688]
Suppose

```
class A(neomodel.StructuredNode):
    z = neomodel.RelationshipTo("B", "TO")

class B(neomodel.StructuredNode):
   pass

a = A().save()
b = B(().save()

# a.z.connect(b)

print(a.z)
```

Notice here that ``a,b`` remained unconnected. Therefore, `a.z` is not going to return something of type `B`.

The way things are setup right now, the above snippet returns `RelationshipDefinition` for 4.0.10 or `RelationshipTo` for 5.0.0 but does not complain that this traversal cannot really "happen" because at the moment, there is nothing on the other side of the relationship. For this to be inferred, we would have to check what `a.z` returns to decide if the relationship is initialised or not.

At the moment [we do not have a relevant exception for this](https://neomodel.readthedocs.io/en/latest/module_documentation.html#module-neomodel.exceptions). This could raise `RelationshipUninitialised`.

**Comment by mariusconjeaud** [comment]
Interesting. So this is not strictly related to the change made to `RelationshipTo/From` in version 5.0.0, right ? This would also be an issue in version 4.0.10 where we only create the generic `RelationshipDefinition` ?

**Comment by aanastasiou** [comment]
Yes, that is correct, I am mentioning the behaviour in both versions as I took the time to check them both. There was no suggestion that this is a consequence of recent changes.

**Comment by mariusconjeaud** [comment]
Hmmmm I tried reproducing this, but when a and b are actually connected for me. See on the left side, a.z.all() returns an object of type B :

<img width="734" alt="image" src="https://user-images.githubusercontent.com/7679761/227168839-ec0d8121-d051-4853-9643-0e62d00769fc.png">

**Comment by aanastasiou** [comment]
@mariusconjeaud Yes, that is correct. The issue is raised for the case when `a` and `b` are not connected. 

And if I try to re-express this: The cardinality of `a->b` is `0..*`.

When *there are* elements "across" `a.z` then we could even do something like `len(a.z)` and get a non-zero integer denoting the number of elements "across" `z`. This part of the behaviour is absolutely fine.

But, when there are zero connections "across" `a.z` then, instead of getting something that is "node-like" or "node-queryable" (i.e. something that we could at least call an `.all()` and get a zero length list), we get a `RelationshipDefinition` (or `RelationshipTo`). This part of the behaviour is the one I propose we should take care of. Whether via an exception or other means.

What do you think?

**Comment by mariusconjeaud** [comment]
Hmmm... So I see your point and why that would be a problem.

BUT, I cannot seem to reproduce it to investigate, because I get something of the same type regardless if I have something on the other side of the relationship or not :

<img width="855" alt="image" src="https://user-images.githubusercontent.com/7679761/227211788-eca6d2fb-7a51-4b7c-b83b-3d892baf1fce.png">

**Comment by aanastasiou** [comment]
I am too right now, running my example from above. Thank you for taking the time to look into it. Let's park it for the moment. I will keep an eye on this.

**Comment by aanastasiou** [comment]
@mariusconjeaud All clear Marius, this was on me. My actual problem was triggered by `inflate()`. When I inflated SPECIFIC nodes, their relationships where not re-instantiated properly. Having reviewed a few (they are too many anyway), I concluded that it was because there was nothing on the other side of the relationship. 

I created the test case. Now, this might be something to keep in mind, I do remember trying this yesterday before I post it and it did have this same behaviour. (I hope it was not because I had already try quite a few things to understand why this error was occuring and I picked the wrong variable when confirming the returned values).

In any case, the bug is not in neomodel, but in my code. Specifically, it was a call that had been indented under the preceding `if` when it should not be. This call was towards the `__init__` of the `super()` and this resulted in the relationships not being re-instated for those specific nodes.

I am sorry for having wasted any of your time.
